tools/clang/tools/dxildll/dxildll.cpp

```diff
@@ -0,0 +1,143 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// dxildll.cpp                                                                //
+// Copyright (C) Microsoft. All rights reserved.                             //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Implements the DxcCreateInstanceand DllMain functions                     //
+// for the dxil.dll module.                                                  //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/Support/Global.h"
+#include "dxc/Support/WinIncludes.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/ManagedStatic.h"
+#include <algorithm>
+#ifdef _WIN32
+#include "Tracing/DxcRuntimeEtw.h"
+#include "dxc/Tracing/dxcetw.h"
+#endif
+
+#ifdef _WIN32
+#define DXC_API_IMPORT
+#else
+#define DXC_API_IMPORT __attribute__((visibility("default")))
+#endif
+
+#include "dxc/dxcisense.h"
+#include "dxc/dxctools.h"
+
+HRESULT CreateDxcValidator(REFIID, LPVOID *);
+HRESULT CreateDxcHashingContainerBuilder(REFIID RRID, LPVOID *V);
+
+// C++ exception specification ignored except to indicate a function is not
+// __declspec(nothrow)
+static HRESULT InitMaybeFail() throw() {
+  bool MemSetup = false;
+  HRESULT HR = (DxcInitThreadMalloc());
+  if (DXC_FAILED(HR))
+    goto Cleanup;
+  DxcSetThreadMallocToDefault();
+  MemSetup = true;
+  if (::llvm::sys::fs::SetupPerThreadFileSystem()) {
+    HR = E_FAIL;
+    goto Cleanup;
+  }
+Cleanup:
```

Comment:
nit: I know it's not your code originally, but this seems a flimsy excuse to use a goto as one of them just falls through anyway and the other could very easily be a conditional block
```suggestion
  if (!DXC_FAILED(HR)) {
    DxcSetThreadMallocToDefault();
    MemSetup = true;
    if (::llvm::sys::fs::SetupPerThreadFileSystem())
      HR = E_FAIL;
  }
```

Reply:
Updated.